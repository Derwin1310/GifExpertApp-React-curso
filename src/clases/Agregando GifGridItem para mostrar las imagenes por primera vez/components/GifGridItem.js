import React from 'react'

//Aqui estamos usando de manera desestructurada las propiedades que le asignamos en GifGrid a GGI, y usamos el titulo y la url para mostrarlos en una etiqueta img en la ol que tenemos en GifGrid
export const GifGridItem = ({title, url}) => {
    console.log(title,url)
  return (
    <div>
        <img src={url} alt={title} />
        <p>{title}</p>
    </div>
  )
}
