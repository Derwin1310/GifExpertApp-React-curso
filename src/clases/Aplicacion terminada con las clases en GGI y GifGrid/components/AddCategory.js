import React, {useState} from 'react';
import PropTypes from 'prop-types';

export const AddCategory = ({setCategories}) => {
	const [inputValue, setinputValue] = useState('Search something!');

	const handleInputValue = e => {
		setinputValue(e.target.value);
	};

	const handleSubmit = e => {
		e.preventDefault();
		if (inputValue.length > 2) {
			setCategories(cats => [inputValue, ...cats]);
			setinputValue('');
		}
	};

	return (
		<form onSubmit={handleSubmit}>
			<input type='text' value={inputValue} onChange={handleInputValue} />
		</form>
	);
};

AddCategory.propTypes = {
	setCategories: PropTypes.func.isRequired,
};
