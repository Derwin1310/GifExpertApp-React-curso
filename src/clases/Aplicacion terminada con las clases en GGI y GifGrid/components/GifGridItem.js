import React from 'react'

export const GifGridItem = ({title, url}) => {
  return (
    //Aqui estamos colocando las animaciones de css de la pagina Animate.css, y la insertamos linkeandola en el html de public, en la pagina estan las instrucciones para aplicar los estilos.
    //Aqui le estamos aplicando un estilo a las imagenes que se muestran en la aplicacion.
    <div className='card animate__animated animate__backInLeft animate__delay-2s'>
        <img src={url} alt={title} />
        <p>{title}</p>
    </div>
  )
}
