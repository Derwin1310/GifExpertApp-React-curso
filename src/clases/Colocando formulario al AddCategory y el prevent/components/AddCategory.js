import React, { useState } from 'react'

export const AddCategory = () => {

    const [inputValue, setinputValue] = useState('Hola mundo');

    //Esta es la funcion que nos permite saber que tecla estamos escribiendo para poder modificar el value del input con el onChange que contiene
    const handleInputValue = (e) => {
        setinputValue(e.target.value);
    };

    //Esta otra funcion nos permite mostrar un mensaje luego de que le damos el submit al formulario y evitar que se refresque colocando el preventDefault().
    const handleSubmit =(e) => {
        e.preventDefault();
        console.log('Submit realizado');
        console.log(inputValue);
    };

    return (
        //Aqui estamos quitando el <> fragment y agregando el form ya que al final hace la misma funcion de mostrar una parte de codigo html en un contenedor, en este caso un formulario
        //El onSubmit nos permite hacer que cuando escribamos algo en el form y le demos a enter la pagina funcione como un boton de submit, aqui le agregamos la funcion de handleSubmit para que no se pueda refrescar y nos muestre un mensaje por la consola.
        <form onSubmit={handleSubmit}>
            <input
                type="text"
                value={inputValue}
                //El onChange lo tenemos que utilizar en los input cuando le colocamos un value por defecto para luego poder escribir en el o modificar el input que contiene por defecto.
                //Le estamos agregando el handleInputValue que lo que hace es leer el valor de la tecla que estamos escribiendo y nos permite modificarlo 
                onChange={handleInputValue}
            />
        </form>
    )
}

