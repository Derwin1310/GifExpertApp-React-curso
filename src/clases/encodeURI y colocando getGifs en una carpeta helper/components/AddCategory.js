import React, {useState} from 'react';
import PropTypes from 'prop-types';

//Aqui estamos agarrando el parametro que le colocamos en el anterior componente como si fuera de el.
export const AddCategory = ({setCategories}) => {
	//Si no le colocamos un valor al parametro del useState, y utilizamos el onChange para el input, nos va a dar un error ya que el valor que se va a cambiar es undefined, y eso daria error.
	//Si no queremos mostrar nada en el input simplemente colocamos un string vacio y ahi no tendra errores porque un input vacio no es igual que undefined.
	const [inputValue, setinputValue] = useState('');

	const handleInputValue = e => {
		setinputValue(e.target.value);
	};

	const handleSubmit = e => {
		e.preventDefault();
		// console.log('Submit realizado');
		// console.log(inputValue);
		//Aqui hacemos una validacion para que el muestre el inputValue solo si es mayor de 2 caracteres y que luego de eso el input sea un string vacio.
		if (inputValue.length > 2) {
			setCategories(cats => [inputValue, ...cats]);
			setinputValue('');
		}
	};

	return (
		<form onSubmit={handleSubmit}>
			<input type='text' value={inputValue} onChange={handleInputValue} />
		</form>
	);
};

//Aqui hacemos que sea un prototipo obligatorio, y al ser una funcion le colocamos PropTypes.func ya que es el que se utiliza para funciones
AddCategory.propTypes = {
	setCategories: PropTypes.func.isRequired,
};
