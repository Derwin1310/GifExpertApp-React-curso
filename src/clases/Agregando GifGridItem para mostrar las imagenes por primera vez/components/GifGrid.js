import React, { useEffect, useState } from 'react'
import { GifGridItem } from './GifGridItem';

export const GifGrid = ({ category }) => {

    //Aqui usamos el useState de esta funcion para poder imprimirla mas abajo con los gif en el li, y lo hacemos un array vacio para que tenga donde almacenar las imagenes
    const [images, setImages] = useState([])

    useEffect(() => {
        getGifs();
    }, [])

    const getGifs = async () => {
        const url = 'https://api.giphy.com/v1/gifs/search?q=Rick and Morty&limit=12&api_key=46Sat15Rc1PEzzKY0bFdQs2ly07cuhtk';
        const res = await fetch(url);
        const { data } = await res.json();

        const gif = data.map(img => {
            return {
                id: img.id,
                title: img.title,
                url: img.images?.downsized_medium.url,
            };
        });
        console.log(gif)
        //Aqui es donde le colocamos los gif al setImages para que al momento de ejecutar la funcion se muestren las imagenes
        setImages(gif)
    };


    return (
        <div>
            <h3>{category}</h3>
            <ol>
                {
                    images.map(img => (
                        <GifGridItem
                            key={img.id}
                            //Aqui le asignamos el operador spread a img para que se traiga todo lo que recorre el images.map que son basicamente las imagenes de getGifs y no tener que colocar el url, title y esas cosas de manera manual, en pocas palabras le estamos asignando como parametro los id, titles y url de las imagenes a GGI.
                            {...img}
                        />
                    ))
                }
            </ol>
        </div>
    )
}
