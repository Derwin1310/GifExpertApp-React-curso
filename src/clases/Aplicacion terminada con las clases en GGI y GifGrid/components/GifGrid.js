import React from 'react';
import {useFetchGifs} from '../hooks/useFetchGifs';
import {GifGridItem} from './GifGridItem';

export const GifGrid = ({category}) => {
	const {data: imgs, loading} = useFetchGifs(category);

	return (
		<>
            {/* Aqui mostramos los estilos al h3 que muestra lo que tiene category cuando insertamos algo en el input y al loading mientras cargan con el setTimeOut.*/}
			<h3 className='card animate__animated animate__bounceInRight animate__delay-5s'>
				{category}
			</h3>
			{loading && (
				<h4 className='animate__animated animate__flash'>Loading...</h4>
			)}
			<div className='card-grid'>
				{imgs.map(img => (
					<GifGridItem key={img.id} {...img} />
				))}
			</div>
		</>
	);
};
