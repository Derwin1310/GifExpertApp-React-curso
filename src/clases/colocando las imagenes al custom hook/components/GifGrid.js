import React from 'react';
import { useFetchGifs } from '../hooks/useFetchGifs';
import {GifGridItem} from './GifGridItem';

export const GifGrid = ({category}) => {
	
    //Aqui lo que hacemos es aplicar el custom hook pero ya utilizandolo de verdad con la peticion https del fetch y le colocamos el category para que muestre las imagenes.
    //De modo que ya no necesitamos el getGifs
    const {data: imgs, loading} = useFetchGifs(category);
    
	return (
		<>
			<h3>{category}</h3>
            {/* Aqui colocamos el loading con una condicion para que muestre un parrafo mientras vienen las imagenes */}
            {loading && <h4>Loading...</h4>}
			<div className='card-grid'>
                {/* Y aqui le hacemos el .map al imgs ya que el es quien tiene la funcion del getGifs para traer las imagenes */}
				{imgs.map(img => (
					<GifGridItem key={img.id} {...img} />
				))}
			</div>
		</>
	);
};
