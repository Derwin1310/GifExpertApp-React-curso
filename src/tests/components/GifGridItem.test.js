import {shallow} from 'enzyme';
import {GifGridItem} from '../../components/GifGridItem';

describe('Tests in Gif Grid Item', () => {
	const title = 'Very nice title';
	const url = 'https://localhost/something.jpg';
	const wrapper = shallow(<GifGridItem title={title} url={url} />);

	test('should show the component correctly', () => {
		expect(wrapper).toMatchSnapshot();
	});

	test('should have a paragraph with the title', () => {
		const p = wrapper.find('p').text();

		expect(p).toBe(title);
	});

	test('should have the same image as the url with props', () => {
		const img = wrapper.find('img');
		// const imgSrc = wrapper.find('img').prop('src');
		// const imgAlt = wrapper.find('img').prop('alt');

		expect(img.prop('src')).toBe(url);
		expect(img.prop('alt')).toBe(title);
	});

	test('should have the className animate_backInLeft', () => {
		const div = wrapper.find('div').hasClass('animate__backInLeft')
		// const div = wrapper.find('div').prop('className').includes('animate__backInLeft')

		// console.log(div)

		// const clas = 'animate_backInLeft';

		expect(div).toBe(true);
	});
});
