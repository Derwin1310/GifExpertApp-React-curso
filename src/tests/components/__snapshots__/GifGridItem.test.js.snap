// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tests in Gif Grid Item should show the component correctly 1`] = `
<div
  className="card animate__animated animate__backInLeft animate__delay-2s"
>
  <img
    alt="Very nice title"
    src="https://localhost/something.jpg"
  />
  <p>
    Very nice title
  </p>
</div>
`;
